model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java

```diff
@@ -17,389 +17,113 @@
-            this.newValue = newValue;
-        }
-
-        @Override
-        public void execute() {
-            oldValue.execute();
-            newValue.execute();
-        }
-
-        @Override
-        public V getValue() {
-            return newValue.getValue();
-        }
-
-        @Override
-        public MapOperation getOperation() {
-            return null;
-        }
-
-        @Override
-        public boolean containsCreate() {
-            return oldValue.containsCreate() || newValue.containsCreate();
-        }
-
-        @Override
-        public boolean containsRemove() {
-            return oldValue.containsRemove() || newValue.containsRemove();
-        }
-
-        @Override
-        public boolean isReplace() {
-            return (newValue.getOperation() == MapOperation.CREATE && oldValue.containsRemove()) ||
-              (oldValue instanceof MapKeycloakTransaction.MapTaskCompose && ((MapTaskCompose) oldValue).isReplace());
-        }
-    }
-
-    private class CreateOperation extends MapTaskWithValue {
-        private final K key;
-
-        public CreateOperation(K key, V value) {
-            super(value);
-            this.key = key;
-        }
-
-        @Override public void execute() { map.create(key, getValue()); }
-        @Override public MapOperation getOperation() { return MapOperation.CREATE; }
-    }
-
-    private class UpdateOperation extends MapTaskWithValue {
-        private final K key;
-
-        public UpdateOperation(K key, V value) {
-            super(value);
-            this.key = key;
-        }
-
-        @Override public void execute() { map.update(key, getValue()); }
-        @Override public MapOperation getOperation() { return MapOperation.UPDATE; }
-    }
-
-    private class DeleteOperation extends MapTaskWithValue {
-        private final K key;
-
-        public DeleteOperation(K key) {
-            super(null);
-            this.key = key;
-        }
+    /**
+     * Instructs this transaction to add a new value into the underlying store on commit.
+     *
+     * @param key an identifier that will be associated with the {@code value}
+     * @param value the value
+     */
+    void create(K key, V value);
 
-        @Override public void execute() { map.delete(key); }
-        @Override public MapOperation getOperation() { return MapOperation.DELETE; }
+    /**
+     * Instructs this transaction to update the {@code value} associated with the identifier {@code key} in the
+     * underlying store on commit, when {@code shouldPut} predicate returns true.
+     *
+     * If the underlying store handles transactions on its own, {@code shouldPut} can be ignored
+     *
+     * @param key identifier of the {@code value}
+     * @param value new version of the value
+     * @param shouldPut predicate to check in commit phase
+     */
+    default void updateIfChanged(K key, V value, Predicate<V> shouldPut) {
+        update(key, value);
     }
```

Comment:
Should the default implementation exist in this case? This method must be adapted to the needs of the particular store and the store implementor must think of the best way to implement this.